File: src/js/workspace/agent.js
// The Claude Code pane: a status strip over a terminal running the CLI in the project directory.
//
// The strip is the point. A Claude session that starts in the right folder but without the MCP
// server wired is not obviously broken - it answers, fluently, out of training data, about a
// library whose API changed underneath it. The four facts across the top are the ones that decide
// whether the answers will be about this project: the CLI is there and which version, the Catalyst
// MCP server is wired, CLAUDE.md exists, and the folder is in the registry the server reads.
//
// When the CLI is not there the pane says where the app looked instead of opening a terminal that
// prints one error and sits dead. On this machine that is the normal case, not the exotic one:
// `claude` ships inside the desktop app and is not on PATH.

import { mountTerminal } from "./terminal.js";

/**
 * Accept an `AgentStatus` in either spelling and give back one shape.
 *
 * serde serialises the struct's fields as they are written unless the Rust side adds
 * `rename_all = "camelCase"`, and which of those happens is not this module's decision to make.
 * Reading both is four lines; guessing wrong is a status strip that reports everything missing on
 * a machine where everything is present.
 */
export function normalizeStatus(raw) {
  const s = raw && typeof raw === "object" ? raw : {};
  const pick = (snake, camel) => (s[snake] !== undefined ? s[snake] : s[camel]);
  const str = (v) => (typeof v === "string" && v.trim() ? v.trim() : null);
  return {
    cli_path: str(pick("cli_path", "cliPath")),
    cli_version: str(pick("cli_version", "cliVersion")),
    mcp_wired: !!pick("mcp_wired", "mcpWired"),
    claude_md: !!pick("claude_md", "claudeMd"),
    project_registered: !!pick("project_registered", "projectRegistered"),
  };
}

/** Everything the agent needs is in place. */
export function isReady(status) {
  const s = normalizeStatus(status);
  return !!(s.cli_path && s.mcp_wired && s.claude_md && s.project_registered);
}

/** The four facts, as chips. */
export function statusChips(status) {
  const s = normalizeStatus(status);
  return [
    {
      id: "cli",
      label: "Claude Code",
      value: s.cli_path ? (s.cli_version || "found") : "not found",
      tone: s.cli_path ? "ok" : "bad",
      detail: s.cli_path || null,
    },
    {
      id: "mcp",
      label: "Catalyst MCP",
      value: s.mcp_wired ? "wired" : "not wired",
      tone: s.mcp_wired ? "ok" : "warn",
      detail: s.mcp_wired ? ".mcp.json names the catalyst server" : "the agent would answer from memory, not from this library",
    },
    {
      id: "claudemd",
      label: "CLAUDE.md",
      value: s.claude_md ? "present" : "missing",
      tone: s.claude_md ? "ok" : "warn",
      detail: s.claude_md ? null : "no project instructions for the session",
    },
    {
      id: "registered",
      label: "Project",
      value: s.project_registered ? "registered" : "not registered",
      tone: s.project_registered ? "ok" : "warn",
      detail: s.project_registered ? null : "the MCP server refuses to write outside the registry",
    },
  ];
}

/**
 * Where the app looks for the CLI, in the order it looks.
 *
 * Taken from the resolution order the backend implements. It is written out because "Claude Code
 * was not found" with no list is unactionable, and the fix - the environment variable - is the
 * first item.
 */
export function cliSearchPlaces() {
  return [
    "the CATALYST_CLAUDE_EXE environment variable",
    "claude / claude.exe on PATH",
    "the newest version folder under %APPDATA%\\Claude\\claude-code\\",
  ];
}

/**
 * What `agent_prepare` actually changed, in one sentence.
 *
 * Prepare is idempotent and mostly silent, so without this the button looks like it did nothing
 * the second time it is pressed - and the case worth reporting is the quiet one: a CLAUDE.md that
 * was already there is left alone on purpose, not overwritten.
 */
export function prepareSummary(before, after) {
  const b = normalizeStatus(before);
  const a = normalizeStatus(after);
  const did = [];
  if (!b.mcp_wired && a.mcp_wired) did.push("wired the Catalyst MCP server");
  else if (a.mcp_wired) did.push("the MCP server was already wired");
  if (!b.claude_md && a.claude_md) did.push("wrote CLAUDE.md");
  else if (a.claude_md) did.push("left the existing CLAUDE.md alone");
  if (!a.mcp_wired) did.push("could not wire the MCP server");
  if (!a.claude_md) did.push("CLAUDE.md is still missing");
  if (!a.project_registered) did.push("this folder is not in the project registry, so the agent may read it but not write to it");
  if (!did.length) return "Nothing to change.";
  return did.join("; ") + ".";
}

// ---------------------------------------------------------------- the pane

const tauri = () => (typeof window === "undefined" ? null : window.__TAURI__ || null);
const invoke = (cmd, args) => {
  const t = tauri();
  if (!t) return Promise.reject(new Error("not running inside the Catalyst app"));
  return t.core.invoke(cmd, args);
};
const messageOf = (err) =>
  typeof err === "string" ? err : (err && err.message) || String(err || "something went wrong");

/**
 * Mount the Claude Code pane.
 *
 * @param {object}  opts
 * @param {Element} opts.el   the element to fill
 * @param {string}  opts.dir  the project directory the session runs in
 * @returns {{ destroy, refresh, prepare, status }}
 */
export function mountAgent({ el, dir }) {
  el.classList.add("ws-agent");
  el.innerHTML = "";

  const strip = document.createElement("div");
  strip.className = "ws-agent-strip";
  const chips = document.createElement("div");
  chips.className = "ws-chips";
  const actions = document.createElement("div");
  actions.className = "ws-agent-actions";
  const prepareBtn = document.createElement("button");
  prepareBtn.type = "button";
  prepareBtn.className = "cat-btn cat-btn--primary";
  prepareBtn.textContent = "Set this project up";
  const recheckBtn = document.createElement("button");
  recheckBtn.type = "button";
  recheckBtn.className = "cat-btn";
  recheckBtn.textContent = "Re-check";
  actions.append(prepareBtn, recheckBtn);
  strip.append(chips, actions);

  const msg = document.createElement("div");
  msg.className = "ws-agent-msg hidden";

  const body = document.createElement("div");
  body.className = "ws-agent-body";

  el.append(strip, msg, body);

  let status = normalizeStatus(null);
  let terminal = null;
  let busy = false;
  let destroyed = false;

  const say = (tone, text) => {
    if (!text) {
      msg.classList.add("hidden");
      msg.textContent = "";
      return;
    }
    msg.classList.remove("hidden");
    msg.dataset.tone = tone;
    msg.textContent = text;
  };

  const paintChips = () => {
    chips.innerHTML = "";
    for (const chip of statusChips(status)) {
      const node = document.createElement("div");
      node.className = "cat-chip ws-chip";
      node.dataset.tone = chip.tone;
      node.dataset.chip = chip.id;
      if (chip.detail) node.title = chip.detail;
      // The identity gives status a shape as well as a colour, so the strip still reads in
      // greyscale and to someone who cannot tell the red from the green.
      const dot = document.createElement("span");
      dot.className = "cat-dot cat-dot--" + chip.tone;
      const k = document.createElement("span");
      k.className = "ws-chip-k";
      k.textContent = chip.label;
      const v = document.createElement("span");
      v.className = "ws-chip-v";
      v.textContent = chip.value;
      node.append(dot, k, v);
      chips.appendChild(node);
    }
    prepareBtn.textContent = isReady(status) ? "Set up again" : "Set this project up";
    prepareBtn.disabled = busy;
    recheckBtn.disabled = busy;
  };

  /** The pane the CLI's absence deserves: what was looked for, where, and the way out. */
  const paintMissingCli = () => {
    body.innerHTML = "";
    const panel = document.createElement("div");
    panel.className = "ws-agent-missing cat-card";
    const h = document.createElement("div");
    h.className = "cat-card__title";
    h.textContent = "Claude Code is not on this machine — or not where the app looked";
    const p = document.createElement("p");
    p.className = "ws-agent-missing-lead";
    p.textContent = "Catalyst looked in three places, in this order:";
    const list = document.createElement("ol");
    list.className = "ws-agent-places";
    for (const place of cliSearchPlaces()) {
      const li = document.createElement("li");
      li.textContent = place;
      list.appendChild(li);
    }
    const tail = document.createElement("p");
    tail.className = "ws-agent-missing-tail";
    tail.textContent =
      "The CLI ships inside the Claude desktop app and is usually not on PATH. Point "
      + "CATALYST_CLAUDE_EXE at that claude.exe, then re-check — nothing needs reinstalling.";
    panel.append(h, p, list, tail);
    body.appendChild(panel);
  };

  const mountSession = () => {
    body.innerHTML = "";
    const host = document.createElement("div");
    host.className = "ws-agent-term";
    body.appendChild(host);
    terminal = mountTerminal({
      el: host,
      kind: "claude",
      cwd: dir,
      args: [],
      onError: (text) => say("bad", text),
    });
  };

  /**
   * Show the right thing under the strip.
   *
   * A live session is never torn down for a status refresh - re-checking mid-conversation must not
   * cost the conversation. The terminal is only rebuilt when the CLI's presence actually flips.
   */
  const paintBody = async () => {
    const have = !!status.cli_path;
    if (have && terminal) return;
    if (!have && terminal) {
      const old = terminal;
      terminal = null;
      await old.destroy();
    }
    if (have) mountSession();
    else paintMissingCli();
  };

  async function readStatus() {
    try {
      const raw = await invoke("agent_status", { dir });
      status = normalizeStatus(raw);
      return true;
    } catch (err) {
      status = normalizeStatus(null);
      say("bad", "Could not read the agent's status: " + messageOf(err));
      return false;
    }
  }

  async function refresh({ quiet = false } = {}) {
    if (destroyed) return status;
    busy = true;
    paintChips();
    if (!quiet) say("dim", "Checking…");
    const ok = await readStatus();
    busy = false;
    if (destroyed) return status;
    paintChips();
    if (ok) say(null, "");
    await paintBody();
    return status;
  }

  async function prepare() {
    if (destroyed || busy) return status;
    const before = status;
    busy = true;
    paintChips();
    say("dim", "Setting the project up…");
    try {
      const raw = await invoke("agent_prepare", { dir });
      status = normalizeStatus(raw);
      busy = false;
      if (destroyed) return status;
      paintChips();
      say(isReady(status) ? "ok" : "warn", prepareSummary(before, status));
      await paintBody();
    } catch (err) {
      busy = false;
      if (destroyed) return status;
      paintChips();
      // Prepare writes files. A failure here is a real one - a path that is not writable, a
      // .mcp.json that is not valid JSON - and it is shown as it came back rather than summarised.
      say("bad", "Could not set the project up: " + messageOf(err));
    }
    return status;
  }

  prepareBtn.addEventListener("click", () => { prepare().catch(() => {}); });
  recheckBtn.addEventListener("click", () => { refresh().catch(() => {}); });

  paintChips();
  // Every failure inside refresh has a place on the strip, so this is only a guard against a
  // rejection that nothing would otherwise await until destroy().
  const first = refresh({ quiet: true }).catch(() => status);

  return {
    /*
     * Put text into the session's prompt without sending it.
     *
     * The editor's "Ask about this file" writes an `@path` reference here and stops, because the
     * question is the user's to ask. Pressing Return for them would send whatever sentence the app
     * invented on their behalf, to an agent that can change their project.
     */
    async send(text) {
      if (!terminal || !text) return false;
      try {
        await terminal.ready;
        await invoke("pty_write", { id: terminal.id, data: String(text) });
        terminal.focus?.();
        return true;
      } catch (e) {
        say("bad", messageOf(e));
        return false;
      }
    },
    async destroy() {
      destroyed = true;
      try { await first; } catch { /* reported already */ }
      if (terminal) {
        const old = terminal;
        terminal = null;
        await old.destroy();
      }
      el.innerHTML = "";
      el.classList.remove("ws-agent");
    },
    refresh,
    prepare,
    get status() {
      return status;
    },
    /** Resolves once the first status read has finished. */
    ready: () => first,
  };
}

File: src-tauri/src/agent.rs
//! What a Claude Code session needs before it is useful in a robot project.
//!
//! A session started in a project folder knows nothing about Catalyst. It will happily write 1.x
//! code — `SubsystemBase`, `CommandScheduler.getInstance()`, `ChassisSpeeds` — because that is what
//! most of the FRC code in the world looks like, and none of it compiles against Catalyst 2.x. Two
//! files fix that, and this module writes them.
//!
//! `.mcp.json` points the session at the MCP server this app already carries, so the API can be
//! looked up instead of recalled. `CLAUDE.md` says which season this is, how it builds, and that
//! nothing gets pushed unasked.
//!
//! # Merging, not replacing
//!
//! `.mcp.json` is the user's file. A team may already have servers in it, and a workspace feature
//! that replaced the file to add one would take those away without saying so. So the document is
//! read, the `catalyst` entry is set, and everything else is written back untouched — and a file
//! that cannot be parsed is an error rather than something to overwrite. `CLAUDE.md` is stronger
//! still: if it exists at all, it is left exactly as it is. Instructions someone wrote for their own
//! project are not ours to replace.

use serde::Serialize;
use std::path::Path;

#[derive(Serialize)]
pub struct AgentStatus {
    /// Where the Claude Code CLI is, when it was found.
    pub cli_path: Option<String>,
    pub cli_version: Option<String>,
    /// Whether this project's `.mcp.json` names the catalyst server.
    pub mcp_wired: bool,
    /// Whether this project has a `CLAUDE.md`. Its contents are not inspected — someone else's
    /// instructions are still instructions.
    pub claude_md: bool,
    /// Whether the projects registry lists this folder. It is the boundary for writing into it.
    pub project_registered: bool,
}

/// What `agent_prepare` puts in `.mcp.json`, as the contract writes it.
fn catalyst_server(server_path: &str) -> serde_json::Value {
    serde_json::json!({ "command": "node", "args": [server_path] })
}

// ------------------------------------------------------------------ the pure decisions

/// Whether a `.mcp.json` already names a usable catalyst server.
///
/// An entry with no command, or with no server path in its arguments, is a leftover rather than a
/// wiring: reporting it as wired would leave the user looking at a green tick and a session that
/// cannot see the library.
pub fn mcp_wired(existing: &str) -> bool {
    let Ok(doc) = serde_json::from_str::<serde_json::Value>(existing) else {
        return false;
    };
    let Some(server) = doc.get("mcpServers").and_then(|s| s.get("catalyst")) else {
        return false;
    };
    let has_command = server
        .get("command")
        .and_then(|c| c.as_str())
        .is_some_and(|c| !c.trim().is_empty());
    let has_path = server
        .get("args")
        .and_then(|a| a.as_array())
        .and_then(|a| a.first())
        .and_then(|a| a.as_str())
        .is_some_and(|a| !a.trim().is_empty());
    has_command && has_path
}

/// `existing` with the catalyst server added or refreshed, and everything else left alone.
pub fn merged_mcp(existing: Option<&str>, server_path: &str) -> Result<String, String> {
    let mut doc: serde_json::Value = match existing.map(str::trim).filter(|s| !s.is_empty()) {
        None => serde_json::json!({}),
        Some(text) => serde_json::from_str(text).map_err(|e| {
            // Overwriting here would be the one unrecoverable outcome: whatever servers that file
            // named are gone, and the user finds out the next time they need one.
            format!(
                ".mcp.json is there but is not valid JSON ({e}). Fix or move it - Catalyst will not \
                 overwrite a file it cannot read."
            )
        })?,
    };

    let Some(root) = doc.as_object_mut() else {
        return Err(".mcp.json does not hold a JSON object, so there is nothing to merge into.".into());
    };
    let servers = root
        .entry("mcpServers")
        .or_insert_with(|| serde_json::json!({}));
    let Some(servers) = servers.as_object_mut() else {
        return Err("\"mcpServers\" in .mcp.json is not an object, so it was left alone.".into());
    };
    servers.insert("catalyst".into(), catalyst_server(server_path));

    serde_json::to_string_pretty(&doc)
        .map(|s| s + "\n")
        .map_err(|e| e.to_string())
}

/// Whether a string reads as a version number: digits, dots, digits.
pub fn looks_like_version(s: &str) -> bool {
    s.contains('.')
        && s.split('.')
            .all(|part| !part.is_empty() && part.starts_with(|c: char| c.is_ascii_digit()))
}

/// The version out of an installer path like `...\claude-code\2.1.271\claude.exe`.
///
/// Free, where running the CLI to ask costs half a second of a process start on every status read.
pub fn version_from_path(exe: &str) -> Option<String> {
    let name = Path::new(exe).parent()?.file_name()?.to_string_lossy().to_string();
    looks_like_version(&name).then_some(name)
}

/// The version out of whatever `claude --version` printed.
pub fn version_from_output(out: &str) -> Option<String> {
    out.split_whitespace()
        .find(|token| looks_like_version(token))
        .map(str::to_string)
}

/// The `CLAUDE.md` written for a project that has none.
///
/// Short on purpose, and only facts that hold for every Catalyst 2.x project: which season it is,
/// where the API comes from, how it builds, and that nothing is pushed unasked. Nothing here guesses
/// at the robot — a file that describes a swerve drive to a team that built an arm is worse than no
/// file, because the session believes it.
pub fn claude_md(project_name: &str) -> String {
    claude_md_with(project_name, shared_memory_index().as_deref())
}

/// Where this machine keeps the memory Claude writes across sessions.
///
/// That memory lives per project directory, under `~/.claude/projects/<slug>/memory`, and
/// `MEMORY.md` is the index a session loads. A session started inside the robot project gets its own
/// slug and therefore none of it — which is why an agent in this window would know the Catalyst API
/// and nothing about the robot it had been worked on all week.
///
/// `CLAUDE.md` can import another file with `@path`, so this points at the index rather than copying
/// it: one line, in a file the user can read and delete. The most recently written index wins,
/// because that is the project directory actually being worked in.
pub fn shared_memory_index() -> Option<String> {
    let home = std::env::var("USERPROFILE")
        .or_else(|_| std::env::var("HOME"))
        .ok()?;
    let projects = Path::new(&home).join(".claude").join("projects");
    let mut best: Option<(std::time::SystemTime, String)> = None;

    for entry in std::fs::read_dir(projects).ok()?.filter_map(|e| e.ok()) {
        let index = entry.path().join("memory").join("MEMORY.md");
        let Ok(meta) = std::fs::metadata(&index) else { continue };
        let Ok(modified) = meta.modified() else { continue };
        if best.as_ref().is_none_or(|(seen, _)| modified > *seen) {
            best = Some((modified, index.to_string_lossy().to_string()));
        }
    }
    best.map(|(_, path)| path)
}

/// The body, with the memory index handed in so a test does not depend on this machine.
pub fn claude_md_with(project_name: &str, memory_index: Option<&str>) -> String {
    let memory = match memory_index {
        Some(path) => format!(
            "\n\
             ## What has already been worked out\n\
             \n\
             @{path}\n\
             \n\
             That is this machine's Catalyst work history: the decisions, the measurements, and the\n\
             things that turned out not to be true. Read it before re-deriving something, and say so\n\
             when you are contradicting it. Delete this section to work without it.\n"
        ),
        None => String::new(),
    };

    format!(
        "# {project_name}\n\
         \n\
         A WPILib 2027 robot project built against FrcCatalyst 2.x.\n\
         \n\
         ## Look the API up, do not recall it\n\
         \n\
         The `catalyst` MCP server is wired into this project in `.mcp.json`. Use it to find\n\
         Catalyst classes, methods and documentation before writing against them. Catalyst 2.x is a\n\
         WPILib 2027 and Commands v3 library, and its API differs from the 1.x code that dominates\n\
         training data: there is no `SubsystemBase`, no `CommandScheduler.getInstance()`,\n\
         `ChassisSpeeds` is `ChassisVelocities`, and `Timer.getFPGATimestamp()` is\n\
         `Timer.getTimestamp()`. Assume nothing; look it up.\n\
         \n\
         ## Building\n\
         \n\
         `devtools gradle -- build` builds this project with the JDK its branch needs.\n\
         \n\
         ## Ask the graph before you read the source\n\
         \n\
         The `catalyst` server can build a structural knowledge graph of THIS project with\n\
         `catalyst_graph_build` — no model calls, no token cost — and then answer where a symbol\n\
         lives (`catalyst_graph_search`), what it connects to (`catalyst_graph_neighbors`), how two\n\
         things relate (`catalyst_graph_path`) and what a file reaches outside itself\n\
         (`catalyst_graph_file`). Build it once per session and ask it first: grepping a robot\n\
         project for a name you half-remember is the slow way to the same answer. The same tools\n\
         already answer for the Catalyst library itself, whose graph ships with the app.\n\
         \n\
         ## Git\n\
         \n\
         Never push, tag or release unless it was asked for in the current session.\n\
         {memory}"
    )
}

// ------------------------------------------------------------------ the files

/// Write the two files into `dir`, merging the first and never replacing the second.
///
/// Split from the command so it can be tested against real files without an app running: the whole
/// point of this function is what it does to a file that is already there.
pub(crate) fn write_agent_files(dir: &Path, server_path: &str) -> Result<(), String> {
    let mcp = dir.join(".mcp.json");
    let existing = std::fs::read_to_string(&mcp).ok();
    let merged = merged_mcp(existing.as_deref(), server_path)?;
    std::fs::write(&mcp, merged).map_err(|e| format!("cannot write {}: {e}", mcp.display()))?;

    let guide = dir.join("CLAUDE.md");
    if !guide.exists() {
        let name = dir
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_else(|| "Robot project".into());
        std::fs::write(&guide, claude_md(&name))
            .map_err(|e| format!("cannot write {}: {e}", guide.display()))?;
    }
    Ok(())
}

/// Look at a project and at this machine, and report both.
fn status_for(dir: &Path, roots: &[String]) -> AgentStatus {
    let cli_path = crate::pty::resolve_claude(&crate::pty::gather())
        .ok()
        .map(|p| p.program);
    let cli_version = cli_path.as_deref().and_then(cli_version);

    let mcp = std::fs::read_to_string(dir.join(".mcp.json")).unwrap_or_default();

    AgentStatus {
        cli_path,
        cli_version,
        mcp_wired: mcp_wired(&mcp),
        claude_md: dir.join("CLAUDE.md").is_file(),
        project_registered: crate::workspace::is_registered(dir, roots),
    }
}

/// The CLI's version: from its path when the installer's layout gives it away, otherwise by asking.
fn cli_version(exe: &str) -> Option<String> {
    version_from_path(exe).or_else(|| ask_version(exe))
}

fn ask_version(exe: &str) -> Option<String> {
    let mut cmd = std::process::Command::new(exe);
    cmd.arg("--version");
    #[cfg(windows)]
    {
        // CREATE_NO_WINDOW. Without it a console window opens and closes on top of the app every
        // time the workspace asks, which looks exactly like something crashing.
        use std::os::windows::process::CommandExt;
        cmd.creation_flags(0x0800_0000);
    }
    let out = cmd.output().ok()?;
    let printed = String::from_utf8_lossy(&out.stdout).to_string();
    version_from_output(&printed)
}

fn registered_roots(app: &tauri::AppHandle) -> Vec<String> {
    crate::projects::list_projects(app.clone())
        .into_iter()
        .map(|p| p.path)
        .collect()
}

// ------------------------------------------------------------------ commands

/// Is the CLI here, is this project wired, and which version would run.
#[tauri::command]
pub fn agent_status(app: tauri::AppHandle, dir: String) -> AgentStatus {
    status_for(Path::new(&dir), &registered_roots(&app))
}

/// Write `.mcp.json` and `CLAUDE.md`, then report where that left the project.
#[tauri::command]
pub fn agent_prepare(app: tauri::AppHandle, dir: String) -> Result<AgentStatus, String> {
    let path = Path::new(&dir);
    if !path.is_dir() {
        return Err(format!("not a folder: {dir}"));
    }

    // The same boundary the editor writes behind. Preparing a project is still writing two files
    // into it, and there is no reason for that to have a weaker rule than saving a file does.
    let roots = registered_roots(&app);
    if !crate::workspace::is_registered(path, &roots) {
        return Err(format!(
            "refused to write into {dir}: it is not a project Catalyst knows about. \
             Import the project first."
        ));
    }

    // The path the app already computes for its own resources. Re-deriving it here would be a
    // second answer to where the server lives, and one of them would eventually be wrong.
    let server_path = crate::mcp_server_path(app.clone())?;
    write_agent_files(path, &server_path)?;
    Ok(status_for(path, &roots))
}

#[cfg(test)]
#[path = "agent_tests.rs"]
mod tests;
